### PR TITLE
Add extra parameters to FindCommandParser to parse

### LINKED ISSUE #104: Add FilterCommand

### PR DESCRIPTION
**Pull request by jevan13**
## Describe your changes
- Add extra parameters for FindCommandParser.parse()
- Parse using CLI Syntax
- Create separate ContainsKeywordsPredicate classes for simplicity
- Updated Testcases

## Issue ticket number and link
closes #104 

## Checklist before requesting a review
- [X] I have performed a self-review of my code.
- [X] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 1]
```diff
@@ -23,16 +27,16 @@ public class FindCommand extends Command {
             + "Parameters: (all/e/ph) KEYWORD [MORE_KEYWORDS]...\n"
             + "Example: " + COMMAND_WORD + " " + ARGUMENT_WORD + " alice bob charlie";
 
-    private final NameContainsKeywordsPredicate predicate;
+    private final PredicateContainer predicateContainer;
```
Comment (InfinityTwo) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:30]: Perhaps the name declaration here could be changed to SCREAMING_SNAKE_CASE?

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 2]
```diff
@@ -22,24 +34,67 @@ public class FindCommandParser implements Parser<FindCommand> {
...
+                    new DepartmentContainsKeywordsPredicate(departmentKeywords);
+            predicateContainer.addDepartmentContainsKeywordsPredicate(departmentContainsKeywordsPredicate);
+        }
+        if (argMultimap.getValue(PREFIX_ROLE).isPresent()) {
+            List<String> roleKeywords = Arrays.stream(
+                    argMultimap.getAllValues(PREFIX_ROLE).get(0).split("\\s+")).toList();
+            RoleContainsKeywordsPredicate roleContainsKeywordsPredicate =
+                    new RoleContainsKeywordsPredicate(roleKeywords);
+            predicateContainer.addRoleContainsKeywordsPredicate(roleContainsKeywordsPredicate);
+        }
```
Comment (InfinityTwo) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java]: Perhaps lines 56-90 could be abstracted into a function that dynamically handles the differences? The statements inside every if statement look the same and perhaps the methods for predicateContainer could be abstracted into one function inside the class to handle the input types?

File: src/main/java/seedu/address/model/person/DepartmentContainsKeywordsPredicate.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.function.Predicate;
+
+import seedu.address.commons.util.StringUtil;
+import seedu.address.commons.util.ToStringBuilder;
+
+/**
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class DepartmentContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/DepartmentContainsKeywordsPredicate.java:13]: Ditto on naming constants

File: src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.function.Predicate;
+
+import seedu.address.commons.util.StringUtil;
+import seedu.address.commons.util.ToStringBuilder;
+
+/**
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class EmailContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java:13]: Ditto

File: src/main/java/seedu/address/model/person/PhoneContainsKeywordsPredicate.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.function.Predicate;
+
+import seedu.address.commons.util.StringUtil;
+import seedu.address.commons.util.ToStringBuilder;
+
+/**
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class PhoneContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/PhoneContainsKeywordsPredicate.java:13]: Ditto

File: src/main/java/seedu/address/model/person/PredicateContainer.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+ * Contains all predicates for FindCommand
+ */
+public class PredicateContainer {
+    private NameContainsKeywordsPredicate nameContainsKeywordsPredicate;
+    private PhoneContainsKeywordsPredicate phoneContainsKeywordsPredicate;
+    private EmailContainsKeywordsPredicate emailContainsKeywordsPredicate;
+    private DepartmentContainsKeywordsPredicate departmentContainsKeywordsPredicate;
+    private RoleContainsKeywordsPredicate roleContainsKeywordsPredicate;
+
+    public PredicateContainer() {
```
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/PredicateContainer.java]: Perhaps there should be an explanation/JavaDoc why the constructor is empty?

File: src/main/java/seedu/address/model/person/RoleContainsKeywordsPredicate.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.function.Predicate;
+
+import seedu.address.commons.util.StringUtil;
+import seedu.address.commons.util.ToStringBuilder;
+
+/**
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class RoleContainsKeywordsPredicate implements Predicate<Person> {
+    private final List<String> keywords;
```
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/RoleContainsKeywordsPredicate.java:13]: Ditto

File: src/test/java/seedu/address/logic/parser/AddressBookParserTest.java [hunk 8]
```diff
@@ -84,21 +88,44 @@ public void parseCommand_exit() throws Exception {
...
+        String input2 = FindCommand.COMMAND_WORD + " " + FindEmployeeCommand.ARGUMENT_WORD + " " + PREFIX_NAME
+                + keywords.stream().collect(Collectors.joining(" "));
+        FindEmployeeCommand actual2 = (FindEmployeeCommand) parser.parseCommand(input2);
+        assertEquals(expected2, actual2);
+
+        // FindPotentialCommand
+        FindPotentialCommand expected3 = new FindPotentialCommand(predicateContainer);
+        String input3 = FindCommand.COMMAND_WORD + " " + FindPotentialCommand.ARGUMENT_WORD + " " + PREFIX_NAME
+                + keywords.stream().collect(Collectors.joining(" "));
+        FindPotentialCommand actual3 = (FindPotentialCommand) parser.parseCommand(input3);
```
Comment (InfinityTwo) [inline src/test/java/seedu/address/logic/parser/AddressBookParserTest.java:120]: I like the comprehensive testing for each Find class but perhaps instead of appending 1, 2, 3 or 4 to the names of the variable, there could be a better naming convention?

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 9]
```diff
@@ -26,23 +31,89 @@ public void parse_invalidArgs_throwsParseException() {
     }
 
     @Test
-    public void parse_validArgs_returnsFindCommand() {
+    public void parse_validNameArgs_returnsFindCommand() {
         // no leading and trailing whitespaces
+        NameContainsKeywordsPredicate nameContainsKeywordsPredicate =
+                new NameContainsKeywordsPredicate(Arrays.asList("Alice", "Bob"));
```
Comment (InfinityTwo) [inline src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:39]: Any reason why you didn't import the valid person's name and/or substring the name to avoid regressions?

Comment (InfinityTwo) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:39]: The same question goes for their details like address, contact, email, etc.

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 10]
```diff
@@ -26,23 +31,89 @@ public void parse_invalidArgs_throwsParseException() {
...
+                new FindCommand(predicateContainer);
+        assertParseSuccess(parser, "all r/Manager President", expectedFindCommand);
+
+        // multiple whitespaces between keywords
+        assertParseSuccess(parser, "all r/\n Manager \n \t President  \t", expectedFindCommand);
     }
 
     @Test
     public void parse_empty_keywords() {
         assertParseFailure(parser, "all", String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
```
Comment (InfinityTwo) [inline src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:109]: Perhaps you can also import the constant for all, ph and as well?

File: src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        // same values -> returns true
+        DepartmentContainsKeywordsPredicate firstPredicateCopy =
+                new DepartmentContainsKeywordsPredicate(firstPredicateKeywordList);
+        assertTrue(firstPredicate.equals(firstPredicateCopy));
+
+        // different types -> returns false
+        assertFalse(firstPredicate.equals(1));
+
+        // null -> returns false
+        assertFalse(firstPredicate.equals(null));
```
Comment (InfinityTwo) [inline src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java:39]: I like the comprehensive test cases but perhaps you could use more more edge case test cases like null and empty strings here for all the other test cases?

File: src/test/java/seedu/address/model/person/NameContainsKeywordsPredicateTest.java [hunk 12]
```diff
@@ -43,19 +43,19 @@ public void equals() {
     public void test_nameContainsKeywords_returnsTrue() {
         // One keyword
         NameContainsKeywordsPredicate predicate = new NameContainsKeywordsPredicate(Collections.singletonList("Alice"));
-        assertTrue(predicate.test(new PersonBuilder().withName("Alice Bob").withIsEmployee(true).build()));
+        assertTrue(predicate.test(new PersonBuilder().withName("Alice").withIsEmployee(true).build()));
```
Comment (InfinityTwo) [inline src/test/java/seedu/address/model/person/NameContainsKeywordsPredicateTest.java:46]: Ditto for imports

**Review by InfinityTwo — CHANGES_REQUESTED** [review]
I think your PR generally looks comprehensive but could use less hardcoding of existing constants/test persons. I think the PR is good to go once that, merge conflicts and codecov are fixed.

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java (hunk 2 above)
Comment (jevan13) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java]: Updated. FindCommandParser will use a PredicateContainer static method to extract the values out from the argumentMultimap. Add more methods in PredicateContainer to set the Predicate fields using the argumentMultimap.

File: src/main/java/seedu/address/model/person/PredicateContainer.java (hunk 6 above)
Comment (jevan13) [reply src/main/java/seedu/address/model/person/PredicateContainer.java]: Will remove constructor

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java (hunk 9 above)
Comment (jevan13) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:39]: I feel its better for readability, to see the actual inputs, actual and expected results.

File: src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java (hunk 11 above)
Comment (jevan13) [reply src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java:39]: What other test cases?

Comment (InfinityTwo) [reply src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java:39]: Every other test case without null tests

**Review by InfinityTwo — APPROVED** [review]
